oshmem/mca/spml/ucx/spml_ucx.h

```diff
@@ -215,22 +278,31 @@ static inline void mca_spml_ucx_aux_unlock(void)
     }
 }
 
-static inline void mca_spml_ucx_cache_mkey(mca_spml_ucx_ctx_t *ucx_ctx,
+static inline int mca_spml_ucx_cache_mkey(mca_spml_ucx_ctx_t *ucx_ctx,
                                            sshmem_mkey_t *mkey, uint32_t segno, int dst_pe)
 {
     ucp_peer_t *peer;
+    spml_ucx_cached_mkey_t *ucx_cached_mkey;
+    int rc;
 
     peer = &(ucx_ctx->ucp_peers[dst_pe]);
-    mkey_segment_init(&peer->mkeys[segno].super, mkey, segno);
+    rc = mca_spml_ucx_ep_mkey_get(peer, segno, &ucx_cached_mkey);
+    if (OSHMEM_SUCCESS != rc) {
+        SPML_UCX_ERROR("mca_spml_ucx_ep_mkey_get failed");
+        return rc;
+    }
+    mkey_segment_init(&(*ucx_cached_mkey).super, mkey, segno);
+    return OSHMEM_SUCCESS;
 }
 
 static inline spml_ucx_mkey_t * 
 mca_spml_ucx_get_mkey(shmem_ctx_t ctx, int pe, void *va, void **rva, mca_spml_ucx_t* module)
 {
+    ucp_peer_t *peer;
     spml_ucx_cached_mkey_t *mkey;
     mca_spml_ucx_ctx_t *ucx_ctx = (mca_spml_ucx_ctx_t *)ctx;
 
-    mkey = ucx_ctx->ucp_peers[pe].mkeys;
+    mkey = ucx_ctx->ucp_peers[pe].mkeys[0];
```

Comment:
Why it's always 0 index? does it mean it's always segment 0?

Reply:
This mkey field holds the super and key where key is of type spml_ucx_cached_mkey and super is of type mkey_segment_t. mkey_segment_t have super and rva_base. So, after this step the pointer to the 0 th segment is passed to map_segment_find_va which adjusts the offsets and find if the virtual address is in segment. If matched, it returns the corresponding mkey.